In [1]:
# Pueden ignorar esto, pero deben correrlo una vez 
import random

# Índice - Pandas
* DataFrames (ETL - Extract, Transform, Load)
* Funciones de agregación (Group by, Pivot Table)
* Merges (Cruce de información)

# Data Frames y ETLs

Pandas es una librería -código de otros- ampliamente usada por científicos de datos en el mundo porque permite manejar con mucha facilidad datos tabulares en Python. Si tiene forma de tabla, seguramente lo pueden trabajar usando Pandas. Al igual que con los tipos de datos que ya vimos, Pandas tiene su propio tipo de objeto. Este es conocido como *DataFrame*


Un DataFrame es el equivalente a una hoja de cálculo en Excel o en Sheets, pero tiene cosas mucho más chéveres que ya veremos. Por lo pronto, carguemos un conjunto de datos con Pandas y veamos qué sucede

In [1]:
# Primero debemos cargar la librería
import pandas as pd # "pd" es un alias, para no tener que escribir "pandas" cada vez

Luego la forma de usar las funciones de pandas es simplemente: pd.funcion_que_quiero()

## Cargar

In [2]:
# Ahora carguemos los datos.
# Se pueden cargar muchos formatos (xlsx, csv, txt, etc.) acá usaremos csv, pero basta con buscar en internet el tipo
# de formato que necesitan para encontrar cómo subirlo a un DataFrame de Pandas

pd.read_csv(r'data/vuelosNacionales/vuelosNacionales.csv')

# la función read_csv recibe la dirección del archivo csv que quieren leer

NOMBRE AEROPUERTO                CIUDAD DEPARTAMENTO  \
0                             EL ENCANTO            EL ENCANTO     AMAZONAS   
1    LA CHORRERA - VIRGILIO BARCO VARGAS           LA CHORRERA     AMAZONAS   
2                             LA PEDRERA            LA PEDRERA     AMAZONAS   
3           LETICIA-ALFREDO VASQUEZ COBO               LETICIA     AMAZONAS   
4                               TARAPACA               LETICIA     AMAZONAS   
..                                   ...                   ...          ...   
792                        SANTA BARBARA               CUMARAL         META   
793                     YATAROS DE MELUA          PUERTO LOPEZ         META   
794                     LA CABA?A - META  SAN CARLOS DE GUAROA         META   
795                       AEROSPORT.C.A.               FLANDES       TOLIMA   
796               INGENIO MAYAGUEZ (C.A)            CANDELARIA        VALLE   

    Pasajeros Salidos - 2019 Pasajeros Llegados - 2019  \
0                         62                        55   
1                        788                       859   
2                      2.517                     2.570   
3                    166.039                   167.011   
4                        758                       701   
..                       ...                       ...   
792                      NaN                       NaN   
793                      NaN                       NaN   
794                      NaN                       NaN   
795                      NaN                       NaN   
796                      NaN                       NaN   

    Carga y Correo Salida (Kg) - 2019 Carga y Correo Llegados (kg) - 2019  \
0                                 805                               2.281   
1                              85.783                             105.966   
2                             339.960                             336.699   
3                           9.465.876                           7.424.085   
4                              26.654                              29.652   
..                                ...                                 ...   
792                               NaN                                 NaN   
793                               NaN                                 NaN   
794                               NaN                                 NaN   
795                               NaN                                 NaN   
796                               NaN                                 NaN   

     Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
0                          31.000                         31.000   
1                         136.000                        135.000   
2                         382.000                        383.000   
3                           1.788                          1.791   
4                         122.000                        123.000   
..                            ...                            ...   
792                           NaN                            NaN   
793                           NaN                            NaN   
794                           NaN                            NaN   
795                           NaN                            NaN   
796                           NaN                            NaN   

    Pasajeros Salidos - 2018  ... Carga y Correo Salida (Kg) - 2005  \
0                         87  ...                               NaN   
1                      1.167  ...                            27.801   
2                      2.731  ...                           111.693   
3                    134.304  ...                         8.494.881   
4                      1.164  ...                             6.587   
..                       ...  ...                               ...   
792                      NaN  ...                               NaN   
793                      NaN  ...                               NaN

In [3]:
# Como siempre, podemos asignarlo a una variable

df = pd.read_csv(r'data/vuelosNacionales/vuelosNacionales.csv')


In [5]:
df.head()

NOMBRE AEROPUERTO       CIUDAD DEPARTAMENTO  \
0                           EL ENCANTO   EL ENCANTO     AMAZONAS   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA     AMAZONAS   
2                           LA PEDRERA   LA PEDRERA     AMAZONAS   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA     AMAZONAS   
4                             TARAPACA      LETICIA     AMAZONAS   

  Pasajeros Salidos - 2019 Pasajeros Llegados - 2019  \
0                       62                        55   
1                      788                       859   
2                    2.517                     2.570   
3                  166.039                   167.011   
4                      758                       701   

  Carga y Correo Salida (Kg) - 2019 Carga y Correo Llegados (kg) - 2019  \
0                               805                               2.281   
1                            85.783                             105.966   
2                           339.960                             336.699   
3                         9.465.876                           7.424.085   
4                            26.654                              29.652   

   Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
0                        31.000                         31.000   
1                       136.000                        135.000   
2                       382.000                        383.000   
3                         1.788                          1.791   
4                       122.000                        123.000   

  Pasajeros Salidos - 2018  ... Carga y Correo Salida (Kg) - 2005  \
0                       87  ...                               NaN   
1                    1.167  ...                            27.801   
2                    2.731  ...                           111.693   
3                  134.304  ...                         8.494.881   
4                    1.164  ...                             6.587   

  Carga y Correo Llegados (kg) - 2005 Número Vuelos Salidos - 2005  \
0                                 NaN                          NaN   
1                              60.326                      190.000   
2                              97.117                      156.000   
3                           6.507.999                        1.127   
4                               9.118                       47.000   

   Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
0                            NaN                       NaN   
1                        192.000                       433   
2                        173.000                       385   
3                          1.071                    26.861   
4                         55.000                       NaN   

  Pasajeros Llegados - 2004 Carga y Correo Salida (Kg) - 2004  \
0                       NaN                               NaN   
1                       540                             2.117   
2                       487                           141.136   
3                    28.076                         7.715.581   
4                       NaN                               NaN   

  Carga y Correo Llegados (kg) - 2004 Número Vuelos Salidos - 2004  \
0                                 NaN                          NaN   
1                               4.081                         53.0   
2                              94.403                        101.0   
3                           6.094.633                        755.0   
4                                 NaN                          NaN   

   Número Vuelos Llegados - 2004  
0                            NaN  
1                           54.0  
2                           99.0  
3                          749.0  
4                            NaN  

[5 rows x 99 columns]

In [5]:
# Esto es lo mismo que hicimos en el módulo para quitar los puntos
# Lo pueden ignorar, la explicación está en el Módulo 1.
for i in df.columns[3:]:      
    if df[i].dtype != "float64":
        df[i].fillna('0', inplace=True) # Esto es para rellenar los nulos
        df[i] = df[i].str.replace('.','').astype(float) # Esto para quitar los puntos
    else:
        df[i].fillna(0, inplace=True)

In [6]:
# Y para ver qué tiene por dentro podemos usar el comando .head()

df.head() # muestra por default los primeros cinco registros

NOMBRE AEROPUERTO       CIUDAD DEPARTAMENTO  \
0                           EL ENCANTO   EL ENCANTO     AMAZONAS   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA     AMAZONAS   
2                           LA PEDRERA   LA PEDRERA     AMAZONAS   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA     AMAZONAS   
4                             TARAPACA      LETICIA     AMAZONAS   

   Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
0                      62.0                       55.0   
1                     788.0                      859.0   
2                    2517.0                     2570.0   
3                  166039.0                   167011.0   
4                     758.0                      701.0   

   Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
0                              805.0                               2281.0   
1                            85783.0                             105966.0   
2                           339960.0                             336699.0   
3                          9465876.0                            7424085.0   
4                            26654.0                              29652.0   

   Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
0                        31.000                         31.000   
1                       136.000                        135.000   
2                       382.000                        383.000   
3                         1.788                          1.791   
4                       122.000                        123.000   

   Pasajeros Salidos - 2018  ...  Carga y Correo Salida (Kg) - 2005  \
0                      87.0  ...                                0.0   
1                    1167.0  ...                            27801.0   
2                    2731.0  ...                           111693.0   
3                  134304.0  ...                          8494881.0   
4                    1164.0  ...                             6587.0   

   Carga y Correo Llegados (kg) - 2005  Número Vuelos Salidos - 2005  \
0                                  0.0                         0.000   
1                              60326.0                       190.000   
2                              97117.0                       156.000   
3                            6507999.0                         1.127   
4                               9118.0                        47.000   

   Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
0                          0.000                       0.0   
1                        192.000                     433.0   
2                        173.000                     385.0   
3                          1.071                   26861.0   
4                         55.000                       0.0   

   Pasajeros Llegados - 2004  Carga y Correo Salida (Kg) - 2004  \
0                        0.0                                0.0   
1                      540.0                             2117.0   
2                      487.0                           141136.0   
3                    28076.0                          7715581.0   
4                        0.0                                0.0   

   Carga y Correo Llegados (kg) - 2004  Número Vuelos Salidos - 2004  \
0                                  0.0                           0.0   
1                               4081.0                          53.0   
2                              94403.0                         101.0   
3                            6094633.0                         755.0   
4                                  0.0                           0.0   

   Número Vuelos Llegados - 2004  
0                            0.0  
1                           54.0  
2                           99.0  
3                          749.0  
4                            0.0  

[5 rows x 99 columns]

In [7]:
# También existe .tail(), para los últimos 5

df.tail(3) # o lo puedo personalizar metiendo un número en el paréntesis. En este caso, los últimos tres

NOMBRE AEROPUERTO                CIUDAD DEPARTAMENTO  \
794        LA CABA?A - META  SAN CARLOS DE GUAROA         META   
795          AEROSPORT.C.A.               FLANDES       TOLIMA   
796  INGENIO MAYAGUEZ (C.A)            CANDELARIA        VALLE   

     Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
794                       0.0                        0.0   
795                       0.0                        0.0   
796                       0.0                        0.0   

     Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
794                                0.0                                  0.0   
795                                0.0                                  0.0   
796                                0.0                                  0.0   

     Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
794                           0.0                            0.0   
795                           0.0                            0.0   
796                           0.0                            0.0   

     Pasajeros Salidos - 2018  ...  Carga y Correo Salida (Kg) - 2005  \
794                       0.0  ...                                0.0   
795                       0.0  ...                                0.0   
796                       0.0  ...                                0.0   

     Carga y Correo Llegados (kg) - 2005  Número Vuelos Salidos - 2005  \
794                                  0.0                           0.0   
795                                  0.0                           0.0   
796                                  0.0                           0.0   

     Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
794                            0.0                       2.0   
795                            0.0                       2.0   
796                            0.0                       2.0   

     Pasajeros Llegados - 2004  Carga y Correo Salida (Kg) - 2004  \
794                        2.0                                0.0   
795                        2.0                                0.0   
796                        8.0                                0.0   

     Carga y Correo Llegados (kg) - 2004  Número Vuelos Salidos - 2004  \
794                                  0.0                           1.0   
795                                  0.0                           1.0   
796                                  0.0                           1.0   

     Número Vuelos Llegados - 2004  
794                            1.0  
795                            1.0  
796                            2.0  

[3 rows x 99 columns]

In [8]:
# Miren que el objeto df es de *tipo* DataFrame

type(df)

pandas.core.frame.DataFrame

Así como a los strings los podemos manipular usando cierta forma de escribir código (corchetes cuadrados, por ejemplo), los DataFrames tienen sus propios comandos para poder manipularlos. 

Las acciones que puedo hacer con cualquier base de datos se conocen como ETL - Extract (extraer información), Transform (transformar información) o Load (cargar información). La parte del Load ya lo hicimos con read_csv, ahora veamos qué funciones necesito para Extraer y para Transformar DataFrames.

## Extraer

In [9]:
# Empecemos por extraer. En programación siempre puedo usar muchas formas para lograr lo mismo, acá veremos un par.
# Para entrar a una columna uso: data_frame["nombre_de_columna"].
# Por ejemplo:

df["CIUDAD"] # Debo usar exactamente el mismo nombre como aparece en los datos (mayúsculas y todo)

0                EL ENCANTO
1               LA CHORRERA
2                LA PEDRERA
3                   LETICIA
4                   LETICIA
               ...         
792                 CUMARAL
793            PUERTO LOPEZ
794    SAN CARLOS DE GUAROA
795                 FLANDES
796              CANDELARIA
Name: CIUDAD, Length: 797, dtype: object

In [10]:
# Cada columna de un DataFrame es de *tipo* Serie

ciudades = df["CIUDAD"]
type(ciudades)

pandas.core.series.Series

In [11]:
# Las series se parecen mucho a las listas. 
# Por ejemplo, puedo usar la notación de [índice]:

ciudades[0] # este es el primer valor de la columna CIUDAD

'EL ENCANTO'

In [12]:
# O usar slices

ciudades[5:10]

5      AMALFI
6      CAREPA
7    CAUCASIA
8    EL BAGRE
9    MEDELLIN
Name: CIUDAD, dtype: object

In [13]:
# Si quiero acceder a varias columnas al mismo tiempo, debo usar [["columna_1", ..., "columna_n"]]
# Por ejemplo:

columnas = ["CIUDAD", "DEPARTAMENTO"] 
df[columnas]

CIUDAD DEPARTAMENTO
0              EL ENCANTO     AMAZONAS
1             LA CHORRERA     AMAZONAS
2              LA PEDRERA     AMAZONAS
3                 LETICIA     AMAZONAS
4                 LETICIA     AMAZONAS
..                    ...          ...
792               CUMARAL         META
793          PUERTO LOPEZ         META
794  SAN CARLOS DE GUAROA         META
795               FLANDES       TOLIMA
796            CANDELARIA        VALLE

[797 rows x 2 columns]

In [14]:
# Y el resultado ya deja de ser una Serie
type(df[columnas])

pandas.core.frame.DataFrame

In [15]:
# Entonces ya no puedo usar la notación de las listas, porque ya no estoy trabajando con una Serie

df[columnas][0]

KeyError: 0

In [ ]:
# Ya vimos cómo acceder a una columna completa, ahora veamos cómo hacer lo mismo para una fila completa.
# Para eso uso .iloc[posición de la fila que quiero].
# Por ejemplo, la primera fila

df.iloc[0] # Nos trae los valores de todas las columnas de la primera fila

In [16]:
# Veamos la segunda

df.iloc[1]

NOMBRE AEROPUERTO                      LA CHORRERA - VIRGILIO BARCO VARGAS
CIUDAD                                                         LA CHORRERA
DEPARTAMENTO                                                      AMAZONAS
Pasajeros Salidos - 2019                                               788
Pasajeros Llegados - 2019                                              859
                                                      ...                 
Pasajeros Llegados - 2004                                              540
Carga y Correo Salida (Kg) - 2004                                     2117
Carga y Correo Llegados (kg) - 2004                                   4081
Número Vuelos Salidos - 2004                                            53
Número Vuelos Llegados - 2004                                           54
Name: 1, Length: 99, dtype: object

In [17]:
# Pero, a diferencia de las listas, el  DataFrame además tiene un *índice*, que es un identificador 
# asignado una sola vez cuando se cargan los datos y *no depende de la posición*
# Por ejemplo:

df_1 = df.tail().copy()
df_1 # Este es un nuevo DataFrame que solo tiene los cinco últimos registros del df original

NOMBRE AEROPUERTO                CIUDAD DEPARTAMENTO  \
792           SANTA BARBARA               CUMARAL         META   
793        YATAROS DE MELUA          PUERTO LOPEZ         META   
794        LA CABA?A - META  SAN CARLOS DE GUAROA         META   
795          AEROSPORT.C.A.               FLANDES       TOLIMA   
796  INGENIO MAYAGUEZ (C.A)            CANDELARIA        VALLE   

     Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
792                       0.0                        0.0   
793                       0.0                        0.0   
794                       0.0                        0.0   
795                       0.0                        0.0   
796                       0.0                        0.0   

     Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
792                                0.0                                  0.0   
793                                0.0                                  0.0   
794                                0.0                                  0.0   
795                                0.0                                  0.0   
796                                0.0                                  0.0   

     Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
792                           0.0                            0.0   
793                           0.0                            0.0   
794                           0.0                            0.0   
795                           0.0                            0.0   
796                           0.0                            0.0   

     Pasajeros Salidos - 2018  ...  Carga y Correo Salida (Kg) - 2005  \
792                       0.0  ...                                0.0   
793                       0.0  ...                                0.0   
794                       0.0  ...                                0.0   
795                       0.0  ...                                0.0   
796                       0.0  ...                                0.0   

     Carga y Correo Llegados (kg) - 2005  Número Vuelos Salidos - 2005  \
792                                  0.0                           0.0   
793                                  0.0                           0.0   
794                                  0.0                           0.0   
795                                  0.0                           0.0   
796                                  0.0                           0.0   

     Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
792                            0.0                      16.0   
793                            0.0                       2.0   
794                            0.0                       2.0   
795                            0.0                       2.0   
796                            0.0                       2.0   

     Pasajeros Llegados - 2004  Carga y Correo Salida (Kg) - 2004  \
792                        0.0                                0.0   
793                        0.0                                0.0   
794                        2.0                                0.0   
795                        2.0                                0.0   
796                        8.0                                0.0   

     Carga y Correo Llegados (kg) - 2004  Número Vuelos Salidos - 2004  \
792                                  0.0                          12.0   
793                                  0.0                           1.0   
794                                  0.0                           1.0   
795                                  0.0                           1.0   
796                                  0.0                           1.0   

     Número Vuelos Llegados - 2004  
792                            0.0  
793                            0.0  
794                            1.0  
795                            1.0  
796                            2.0  

[5 rows x 99 col

In [18]:
# Si quisiera acceder a la *primera* fila de este DataFrame, puedo usar [0].
# Veamos:

df_1.iloc[0]

NOMBRE AEROPUERTO                      SANTA BARBARA
CIUDAD                                       CUMARAL
DEPARTAMENTO                                    META
Pasajeros Salidos - 2019                           0
Pasajeros Llegados - 2019                          0
                                           ...      
Pasajeros Llegados - 2004                          0
Carga y Correo Salida (Kg) - 2004                  0
Carga y Correo Llegados (kg) - 2004                0
Número Vuelos Salidos - 2004                      12
Número Vuelos Llegados - 2004                      0
Name: 792, Length: 99, dtype: object

In [19]:
# Pero el *índice* de la primera fila ya no es 0 sino 792.
# Miren el número antes de la primera columna:

df_1.head()

NOMBRE AEROPUERTO                CIUDAD DEPARTAMENTO  \
792           SANTA BARBARA               CUMARAL         META   
793        YATAROS DE MELUA          PUERTO LOPEZ         META   
794        LA CABA?A - META  SAN CARLOS DE GUAROA         META   
795          AEROSPORT.C.A.               FLANDES       TOLIMA   
796  INGENIO MAYAGUEZ (C.A)            CANDELARIA        VALLE   

     Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
792                       0.0                        0.0   
793                       0.0                        0.0   
794                       0.0                        0.0   
795                       0.0                        0.0   
796                       0.0                        0.0   

     Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
792                                0.0                                  0.0   
793                                0.0                                  0.0   
794                                0.0                                  0.0   
795                                0.0                                  0.0   
796                                0.0                                  0.0   

     Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
792                           0.0                            0.0   
793                           0.0                            0.0   
794                           0.0                            0.0   
795                           0.0                            0.0   
796                           0.0                            0.0   

     Pasajeros Salidos - 2018  ...  Carga y Correo Salida (Kg) - 2005  \
792                       0.0  ...                                0.0   
793                       0.0  ...                                0.0   
794                       0.0  ...                                0.0   
795                       0.0  ...                                0.0   
796                       0.0  ...                                0.0   

     Carga y Correo Llegados (kg) - 2005  Número Vuelos Salidos - 2005  \
792                                  0.0                           0.0   
793                                  0.0                           0.0   
794                                  0.0                           0.0   
795                                  0.0                           0.0   
796                                  0.0                           0.0   

     Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
792                            0.0                      16.0   
793                            0.0                       2.0   
794                            0.0                       2.0   
795                            0.0                       2.0   
796                            0.0                       2.0   

     Pasajeros Llegados - 2004  Carga y Correo Salida (Kg) - 2004  \
792                        0.0                                0.0   
793                        0.0                                0.0   
794                        2.0                                0.0   
795                        2.0                                0.0   
796                        8.0                                0.0   

     Carga y Correo Llegados (kg) - 2004  Número Vuelos Salidos - 2004  \
792                                  0.0                          12.0   
793                                  0.0                           1.0   
794                                  0.0                           1.0   
795                                  0.0                           1.0   
796                                  0.0                           1.0   

     Número Vuelos Llegados - 2004  
792                            0.0  
793                            0.0  
794                            1.0  
795                            1.0  
796                            2.0  

[5 rows x 99 col

In [20]:
# Antes de seguir, recordemos cómo se ve nuestro DataFrame original

df.head()

NOMBRE AEROPUERTO       CIUDAD DEPARTAMENTO  \
0                           EL ENCANTO   EL ENCANTO     AMAZONAS   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA     AMAZONAS   
2                           LA PEDRERA   LA PEDRERA     AMAZONAS   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA     AMAZONAS   
4                             TARAPACA      LETICIA     AMAZONAS   

   Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
0                      62.0                       55.0   
1                     788.0                      859.0   
2                    2517.0                     2570.0   
3                  166039.0                   167011.0   
4                     758.0                      701.0   

   Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
0                              805.0                               2281.0   
1                            85783.0                             105966.0   
2                           339960.0                             336699.0   
3                          9465876.0                            7424085.0   
4                            26654.0                              29652.0   

   Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
0                        31.000                         31.000   
1                       136.000                        135.000   
2                       382.000                        383.000   
3                         1.788                          1.791   
4                       122.000                        123.000   

   Pasajeros Salidos - 2018  ...  Carga y Correo Salida (Kg) - 2005  \
0                      87.0  ...                                0.0   
1                    1167.0  ...                            27801.0   
2                    2731.0  ...                           111693.0   
3                  134304.0  ...                          8494881.0   
4                    1164.0  ...                             6587.0   

   Carga y Correo Llegados (kg) - 2005  Número Vuelos Salidos - 2005  \
0                                  0.0                         0.000   
1                              60326.0                       190.000   
2                              97117.0                       156.000   
3                            6507999.0                         1.127   
4                               9118.0                        47.000   

   Número Vuelos Llegados - 2005  Pasajeros Salidos - 2004  \
0                          0.000                       0.0   
1                        192.000                     433.0   
2                        173.000                     385.0   
3                          1.071                   26861.0   
4                         55.000                       0.0   

   Pasajeros Llegados - 2004  Carga y Correo Salida (Kg) - 2004  \
0                        0.0                                0.0   
1                      540.0                             2117.0   
2                      487.0                           141136.0   
3                    28076.0                          7715581.0   
4                        0.0                                0.0   

   Carga y Correo Llegados (kg) - 2004  Número Vuelos Salidos - 2004  \
0                                  0.0                           0.0   
1                               4081.0                          53.0   
2                              94403.0                         101.0   
3                            6094633.0                         755.0   
4                                  0.0                           0.0   

   Número Vuelos Llegados - 2004  
0                            0.0  
1                           54.0  
2                           99.0  
3                          749.0  
4                            0.0  

[5 rows x 99 columns]

In [21]:
# Ya que sabemos acceder a toda una columna y a toda una fila, veamos cómo acceder a una intersección de las dos,
# es decir, al valor específico de una celda. Existen dos formas: 1) Primero accedo a la columna y luego escojo la fila
# o 2) Primero escojo la fila y luego la columna.
# Ejemplo del 1):

df["CIUDAD"] # Acá accedo a toda la columna
df["CIUDAD"][3] # Acá accedo a la *cuarta* fila de la columna CIUDAD

'LETICIA'

In [22]:
# Ejemplo del 2):

df.iloc[3] # Acá accedo a toda la cuarta fila
df.iloc[3]["CIUDAD"] # Acá accedo a la *columna* CIUDAD de la cuarta fila

'LETICIA'

## Ejercicio 1
Guarde la columna de pasajeros salidos en el 2018 en una variable que se llame salida_2018. En otra variable llamada la_perd guarde toda la información disponible del aeropuerto LA PEDRERA.

In [23]:
# Acá va su código

salida_2018 = df[] # Acá el nombre de la columna que nos interesa

la_perd = df.iloc[] # Acá la posición del aeropuerto que nos interesa

SyntaxError: invalid syntax (<ipython-input-23-f94b60b65fe8>, line 3)

## Ejercicio 2
Utilizando la información del Ejercicio 1, responda las siguientes preguntas:

* ¿Cuál fue la mayor cantidad de pasajeros que salieron de un aeropuerto colombiano en 2018?, ¿cuál fue la menor? Y, ¿cuál fue el promedio? 

* ¿Cómo cambió la salida de pasajeros en el aeropuerto de La pedrera entre el 2019 y el 2018? ¿Aumentó, se redujo o se mantuvo igual? 

Pistas: para sacar máximos y mínimos pueden usar max(lista) o min(lista). Un promedio se puede calcular como suma/tamaño, pero existen funciones en Python (busquen en internet)

Respuestas:

* Max: 16402228, Min: 0, Avg: 43745.95
* Se redujo en 214 pasajeros

In [24]:
# Acá va su código

## Transformar

Algo que nos gustaría hacer es transformar nuestras tablas y no solo acceder a lo que hay dentro. Hay muchos tipos de transformaciones: podría, por ejemplo, querer cambiar una celda específica, o crear una nueva columna, o ajustar todos los valores de filas que cumplan con alguna condición. Aquí veremos cómo hacer esas transformaciones.

In [25]:
# Empecemos viendo el comando .shape

print(df.shape)

# Nos da las dimensiones de la tabla. El primer valor es el número de filas y el segundo el de columnas

(797, 99)


In [26]:
# Si hago

filas = df.shape[0]

print(filas)

# Me traigo solo el número de filas

797


In [27]:
# Ahora utilicemos eso para crear una nueva columna repleta de ceros.

# Lo primero que hago es crear una lista de ceros que tenga *la misma cantidad de filas que la tabla*

lista_ceros = [0]*filas

print(lista_ceros)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [28]:
# Y por último, se la asigno a la nueva columna que llamaré "nueva_columna":
df["nueva_columna"] = lista_ceros

In [29]:
# ¡Y funcionó! Ya tenemos una nueva columna
df[["nueva_columna"]]

nueva_columna
0                0
1                0
2                0
3                0
4                0
..             ...
792              0
793              0
794              0
795              0
796              0

[797 rows x 1 columns]

Pero una columna con ceros no es para nada interesante... Hagamos algo más con ella. Por ejemplo, cambiemos esta columna a un indicador de la llegada de pasajeros en el 2019. Lo que queremos es tener 0 para los aeropuertos que estén por debajo del promedio del 2019 y 1 para los que estén igual o por encima.

Para eso necesitamos primero saber el promedio del 2019:

In [30]:
# Acá tenemos el promedio. Si lo hicieron arriba de otra forma, utilicen este método para comparar su respuesta

promedio_2019 = df["Pasajeros Llegados - 2019"].mean()
print(promedio_2019)

47483.074027603514


In [31]:
# Ya que tenemos el promedio, usemos este número para marcar los aeropuertos que estén por encima. Para eso, recordemos, 
# vamos a poner un 1 en la columna que acabamos de crear.

# Introduciendo... loc. Este comando se usa para acceder a grupos de celdas y, cuando se necesite, transformarlas.
# Se escribe así df.loc[una condición o un índice, columnas que me quiero traer] = valor a asignar.

# Veamos la primera parte del comando:

df.loc[0, "Pasajeros Llegados - 2019"] # Me trae el primer registro de la columna "Pasajeros Llegados - 2019"

55.0

In [32]:
# También puedo usar condiciones:

df.loc[df["DEPARTAMENTO"]=="AMAZONAS"] # Si no especifico, me trae todas las columnas

NOMBRE AEROPUERTO       CIUDAD DEPARTAMENTO  \
0                           EL ENCANTO   EL ENCANTO     AMAZONAS   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA     AMAZONAS   
2                           LA PEDRERA   LA PEDRERA     AMAZONAS   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA     AMAZONAS   
4                             TARAPACA      LETICIA     AMAZONAS   

   Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
0                      62.0                       55.0   
1                     788.0                      859.0   
2                    2517.0                     2570.0   
3                  166039.0                   167011.0   
4                     758.0                      701.0   

   Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
0                              805.0                               2281.0   
1                            85783.0                             105966.0   
2                           339960.0                             336699.0   
3                          9465876.0                            7424085.0   
4                            26654.0                              29652.0   

   Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
0                        31.000                         31.000   
1                       136.000                        135.000   
2                       382.000                        383.000   
3                         1.788                          1.791   
4                       122.000                        123.000   

   Pasajeros Salidos - 2018  ...  Carga y Correo Llegados (kg) - 2005  \
0                      87.0  ...                                  0.0   
1                    1167.0  ...                              60326.0   
2                    2731.0  ...                              97117.0   
3                  134304.0  ...                            6507999.0   
4                    1164.0  ...                               9118.0   

   Número Vuelos Salidos - 2005  Número Vuelos Llegados - 2005  \
0                         0.000                          0.000   
1                       190.000                        192.000   
2                       156.000                        173.000   
3                         1.127                          1.071   
4                        47.000                         55.000   

   Pasajeros Salidos - 2004  Pasajeros Llegados - 2004  \
0                       0.0                        0.0   
1                     433.0                      540.0   
2                     385.0                      487.0   
3                   26861.0                    28076.0   
4                       0.0                        0.0   

   Carga y Correo Salida (Kg) - 2004  Carga y Correo Llegados (kg) - 2004  \
0                                0.0                                  0.0   
1                             2117.0                               4081.0   
2                           141136.0                              94403.0   
3                          7715581.0                            6094633.0   
4                                0.0                                  0.0   

   Número Vuelos Salidos - 2004  Número Vuelos Llegados - 2004  nueva_columna  
0                           0.0                            0.0              0  
1                          53.0                           54.0              0  
2                         101.0                           99.0              0  
3                         755.0                          749.0              0  
4                           0.0                            0.0              0  

[5 rows x 100 columns]

In [33]:
# Entonces, si quiero traerme la columna con los nombres de todos los aeropuertos en Amazonas, hago:

df.loc[df["DEPARTAMENTO"]=="AMAZONAS", "NOMBRE AEROPUERTO"]

0                             EL ENCANTO
1    LA CHORRERA - VIRGILIO BARCO VARGAS
2                             LA PEDRERA
3           LETICIA-ALFREDO VASQUEZ COBO
4                               TARAPACA
Name: NOMBRE AEROPUERTO, dtype: object

In [34]:
# O si quiero el nombre y la ciudad

columnas=["NOMBRE AEROPUERTO", "CIUDAD"]

df.loc[df["DEPARTAMENTO"]=="AMAZONAS", columnas]

NOMBRE AEROPUERTO       CIUDAD
0                           EL ENCANTO   EL ENCANTO
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA
2                           LA PEDRERA   LA PEDRERA
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA
4                             TARAPACA      LETICIA

In [35]:
# Retomando el ejemplo, traigamos todos los aeropuertos por encima del promedio

## Ejercicio 3

Traiga todas las columnas de los aeropuertos cuyos pasajeros llegados en 2019 están por encima del promedio.

La respuesta estará dos celdas más adelantes, así que intente la suya propia antes de seguir.

In [36]:
# Acá su código

In [37]:
# Para eso usamos loc y el promedio que ya calculamos:
 
df.loc[df["Pasajeros Llegados - 2019"] >= promedio_2019].head() # La respuesta es un DataFrame, así que me traigo solo unos pocos

NOMBRE AEROPUERTO                CIUDAD DEPARTAMENTO  \
3     LETICIA-ALFREDO VASQUEZ COBO               LETICIA     AMAZONAS   
6        ANTONIO ROLDAN BETANCOURT                CAREPA    ANTIOQUIA   
9         MEDELLIN - OLAYA HERRERA              MEDELLIN    ANTIOQUIA   
14      RIONEGRO - JOSE M. CORDOVA  RIONEGRO - ANTIOQUIA    ANTIOQUIA   
17  ARAUCA - SANTIAGO PEREZ QUIROZ    ARAUCA - MUNICIPIO       ARAUCA   

    Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
3                   166039.0                   167011.0   
6                   108134.0                   106839.0   
9                   540537.0                   543474.0   
14                 4615270.0                  4589739.0   
17                   66429.0                    66058.0   

    Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
3                           9465876.0                            7424085.0   
6                              5745.0                              34539.0   
9                            904795.0                             744595.0   
14                         85571822.0                           30050271.0   
17                           638616.0                             780449.0   

    Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
3                          1.788                          1.791   
6                          2.869                          2.872   
9                         21.015                         20.984   
14                        38.018                         37.628   
17                         1.729                          1.729   

    Pasajeros Salidos - 2018  ...  Carga y Correo Llegados (kg) - 2005  \
3                   134304.0  ...                            6507999.0   
6                   103750.0  ...                             516100.0   
9                   578105.0  ...                            2306309.0   
14                 4029910.0  ...                           47391998.0   
17                   60019.0  ...                            1570702.0   

    Número Vuelos Salidos - 2005  Número Vuelos Llegados - 2005  \
3                          1.127                          1.071   
6                          2.612                          2.603   
9                         23.669                         23.533   
14                        17.347                         17.181   
17                         2.689                          2.584   

    Pasajeros Salidos - 2004  Pasajeros Llegados - 2004  \
3                    26861.0                    28076.0   
6                    64444.0                    64506.0   
9                   408158.0                   401311.0   
14                 1024072.0                  1032299.0   
17                   30128.0                    31560.0   

    Carga y Correo Salida (Kg) - 2004  Carga y Correo Llegados (kg) - 2004  \
3                           7715581.0                            6094633.0   
6                            185745.0                             534130.0   
9                           3481824.0                            2092330.0   
14                         75550528.0                           36999608.0   
17                           894853.0                            1709252.0   

    Número Vuelos Salidos - 2004  Número Vuelos Llegados - 2004  nueva_columna  
3                        755.000                        749.000              0  
6                          2.750                          2.745              0  
9                         23.551                         23.451              0  
14                        16.628                         16.674              0  
17                         1.612                          1.607              0  

[5 rows x 100 columns]

In [38]:
# Ahora solo traigamos la columna que queremos cambiar ("nueva_columna"):

df.loc[df["Pasajeros Llegados - 2019"] >= promedio_2019, "nueva_columna"]

3      0
6      0
9      0
14     0
17     0
26     0
29     0
40     0
42     0
47     0
104    0
116    0
126    0
134    0
138    0
147    0
155    0
159    0
195    0
200    0
201    0
202    0
205    0
208    0
211    0
213    0
215    0
217    0
238    0
245    0
308    0
Name: nueva_columna, dtype: int64

In [39]:
# Por último, cambiemos el valor de esa columna por un 1, pues son los que cumplen la condición

df.loc[df["Pasajeros Llegados - 2019"] >= promedio_2019, "nueva_columna"] = 1

In [40]:
# Y veamos cómo queda:
print("El promedio 2019: " + str(promedio_2019))
columnas=["Pasajeros Llegados - 2019", "nueva_columna"]
df[columnas]

El promedio 2019: 47483.074027603514


Pasajeros Llegados - 2019  nueva_columna
0                         55.0              0
1                        859.0              0
2                       2570.0              0
3                     167011.0              1
4                        701.0              0
..                         ...            ...
792                        0.0              0
793                        0.0              0
794                        0.0              0
795                        0.0              0
796                        0.0              0

[797 rows x 2 columns]

¡Funciona! Cuando el valor de pasajeros es menor al promedio tenemos un 0 y cuando es mayor o igual tenemos un 1. Y ya sabemos transformar nuestro DataFrame

# Funciones de agregación

Sigamos con el ejemplo de antes. Ya que tenemos nuestro indicador, ahora un buen dato sería saber *cuántos* aeropuertos están por encima del promedio y cuántos por debajo. Una manera sería contar con la mano, pero vamos a aprovechar este problema para introducir las funciones de agregación. 

Las funciones de agregación, como su nombre lo indica, se usan para agrupar (o agregar) datos y luego operar sobre ellos. Las operaciones pueden ser de todo tipo, como ya veremos. Por lo pronto, entendamos esta tabla dummy:

    CIUDAD  | DEPARTAMENTO | CASOS COVID

    Tumaco  |    Nariño    |    30
    Pasto   |    Nariño    |    20
    Cali    |    Valle     |    100
    Palmira |    Valle     |    50


Si yo quiero saber los casos totales por departamento, debo *agregar* esa variable y luego sumar. Es decir:

agregar(DEPARTAMENTO)  |   sumar(CASOS COVID)

        Nariño         |        50
        Valle          |        150
        

Pero también podría, en lugar de saber el total de casos de COVID, contar el número de ciudades. Es decir: 

agregar(DEPARTAMENTO)  |   contar(CIUDAD)

        Nariño         |        2
        Valle          |        2


Fun fact: lo que acabamos de hacer se llama pivot table. La gente lo usa en Excel a cada rato, pero ya veremos por qué en Python tiene más poder.

Y de eso se tratan las funciones que veremos a continuación: cómo usar Pandas para agrupar datos.
        
    

In [41]:
# Dicho todo lo anterior, agreguemos la nueva_columna y contemos cuántos tienen 1 y cuántos 0
# Para eso usamos groupby, así:

df.groupby(by="nueva_columna").count()

# La sintaxis se explica sola:

# data_frame.groupby(by="nombre de las columna que quiero agrupar").operación()
# pude usar .sum() o alguna función hecha por mí

NOMBRE AEROPUERTO  CIUDAD  DEPARTAMENTO  \
nueva_columna                                            
0                            766     766           766   
1                             31      31            31   

               Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
nueva_columna                                                        
0                                   766                        766   
1                                    31                         31   

               Carga y Correo Salida (Kg) - 2019  \
nueva_columna                                      
0                                            766   
1                                             31   

               Carga y Correo Llegados (kg) - 2019  \
nueva_columna                                        
0                                              766   
1                                               31   

               Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
nueva_columna                                                                
0                                       766                            766   
1                                        31                             31   

               Pasajeros Salidos - 2018  ...  \
nueva_columna                            ...   
0                                   766  ...   
1                                    31  ...   

               Carga y Correo Salida (Kg) - 2005  \
nueva_columna                                      
0                                            766   
1                                             31   

               Carga y Correo Llegados (kg) - 2005  \
nueva_columna                                        
0                                              766   
1                                               31   

               Número Vuelos Salidos - 2005  Número Vuelos Llegados - 2005  \
nueva_columna                                                                
0                                       766                            766   
1                                        31                             31   

               Pasajeros Salidos - 2004  Pasajeros Llegados - 2004  \
nueva_columna                                                        
0                                   766                        766   
1                                    31                         31   

               Carga y Correo Salida (Kg) - 2004  \
nueva_columna                                      
0                                            766   
1                                             31   

               Carga y Correo Llegados (kg) - 2004  \
nueva_columna                                        
0                                              766   
1                                               31   

               Número Vuelos Salidos - 2004  Número Vuelos Llegados - 2004  
nueva_columna                                                               
0                                       766                            766  
1                                        31                             31  

[2 rows x 99 columns]

Lo que nos devuelve es el resultado de la operación para cada columna. Es decir, se fue a la columna NOMBRE AEROPUERTO, agregó utilizando nueva_columna y luego contó registros. Luego se fue a la columna CIUDAD e hizo lo mismo, y así. Por esta razón el resultado es el mismo, no deberían diferenciarse.

La cosa cambia si en vez de contar le diera, por ejemplo, sumar, porque en ese caso cada columna suma para un valor diferente. Solo para que ver qué pasa, hagamos eso:

In [42]:
df.groupby(by="nueva_columna").sum()

Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
nueva_columna                                                        
0                              545699.0                   547732.0   
1                            37505798.0                 37296278.0   

               Carga y Correo Salida (Kg) - 2019  \
nueva_columna                                      
0                                     12445454.0   
1                                    585454496.0   

               Carga y Correo Llegados (kg) - 2019  \
nueva_columna                                        
0                                       16630872.0   
1                                      407723305.0   

               Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
nueva_columna                                                                
0                                 29162.081                      29387.993   
1                                   390.521                        389.562   

               Pasajeros Salidos - 2018  Pasajeros Llegados - 2018  \
nueva_columna                                                        
0                              547835.0                   551004.0   
1                            34317690.0                 34096344.0   

               Carga y Correo Salida (Kg) - 2018  \
nueva_columna                                      
0                                     13889089.0   
1                                    615283120.0   

               Carga y Correo Llegados (kg) - 2018  ...  \
nueva_columna                                       ...   
0                                       18165309.0  ...   
1                                      409282625.0  ...   

               Carga y Correo Salida (Kg) - 2005  \
nueva_columna                                      
0                                     13418665.0   
1                                    518593997.0   

               Carga y Correo Llegados (kg) - 2005  \
nueva_columna                                        
0                                       23264229.0   
1                                      343927192.0   

               Número Vuelos Salidos - 2005  Número Vuelos Llegados - 2005  \
nueva_columna                                                                
0                                 35159.746                       34792.86   
1                                  1601.265                        1671.40   

               Pasajeros Salidos - 2004  Pasajeros Llegados - 2004  \
nueva_columna                                                        
0                              286931.0                   289280.0   
1                            10873845.0                 10806610.0   

               Carga y Correo Salida (Kg) - 2004  \
nueva_columna                                      
0                                      9893845.0   
1                                    498837156.0   

               Carga y Correo Llegados (kg) - 2004  \
nueva_columna                                        
0                                       17565527.0   
1                                      317925984.0   

               Número Vuelos Salidos - 2004  Número Vuelos Llegados - 2004  
nueva_columna                                                               
0                                 31715.823                      31218.593  
1                                  2150.737                       2138.494  

[2 rows x 96 columns]

Como la suma solo tiene sentido para columnas numéricas, ahora solo esas nos aparecen y el resultado cambia. De nuevo, Python se fue columna por columna y sumó los valores de cada columna agrupando por "nueva_columna". ¡Perfecto!

In [43]:
# De regreso a nuestro conteo...
# Como solo necesitamos una columna para el conteo, entonces nos podemos quedar con cualquiera
# Yo voy a escoger CIUDAD, pero puede ser cualquier otra.

df.groupby(by="nueva_columna").count()["CIUDAD"]

# ¡Y listo! Hay 31 igual o por encima y 766 por debajo, ¿qué nos dice eso?

nueva_columna
0    766
1     31
Name: CIUDAD, dtype: int64

In [44]:
# Veamos otra forma de llegar al mismo resultado usando una función llamada pivot_table
# Se usa así:

pd.pivot_table(
               index= "nueva_columna", # Columna por la que quiero agrupar
               aggfunc={"CIUDAD":"count"}, # Qué operación quiero hacer en cada columna cuando agrupe
               data=df # Sobre qué Data Frame haré la operación 
               )

# ¡El resultado es exactamente el mismo! La ventaja de Pivot Table es que puedo hacer una operación diferente para cada 
# columna, mientras que con groupby la operación se hace para todas

CIUDAD
nueva_columna        
0                 766
1                  31

La función pivot_table tiene muchos más parámetros que la hacen una función muy poderosa. Pueden visitar la [documentación](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html) para ver cómo usarlos

## Ejercicio 4

Usando funciones de agregación (la que prefiera), responda las siguientes preguntas:

* ¿Cuántos aeropuertos hay por departamento? Y, ¿cuál es el departamento con más aeropuertos?
* ¿Cuántas ciudades solo tienen un aeropuerto?
* ¿Cómo fue el cambio en la salida de pasajeros entre 2019 y 2018 en cada ciudad?

Pistas: 1) Para organizar la tabla pueden usar la función sort_values(), busquen en internet cómo funciona.

Respuestas:

* Departamento con más aeropuertos: CASANARE
* Hay 173 ciudades con un solo aeropuerto


In [45]:
# Acá va su código

# Merges

Lo último que veremos hoy, son los cruces o merges. Con esta posibilidad, Pandas queda listo para que hagamos maravillas.

Muchas veces tendremos información en diferentes tablas. Por ejemplo, sigamos con el ejemplo del COVID de las funciones de agregación y veamos cómo podríamos tener información regada que nos gustaría tener en el mismo lugar:

    CIUDAD  | DEPARTAMENTO | CASOS COVID

    Tumaco  |    Nariño    |    30
    Pasto   |    Nariño    |    20
    Cali    |    Valle     |    100
    Palmira |    Valle     |    50
    

Podríamos tener otra tabla con las vacunas:

    CIUDAD  | DEPARTAMENTO |  VACUNAS

    Tumaco  |    Nariño    |    13
    Pasto   |    Nariño    |    18
    Cali    |    Valle     |    70
    Palmira |    Valle     |    20
    

Pero tener esa información regada no es algo que nos resulte muy cómodo para poder trabajar. Lo ideal sería tener una tabla así:

    CIUDAD  | DEPARTAMENTO | CASOS COVID |  VACUNAS

    Tumaco  |    Nariño    |    30       |    13
    Pasto   |    Nariño    |    20       |    18
    Cali    |    Valle     |    100      |    70
    Palmira |    Valle     |    50       |    20
    
Para poder unificar esa información, existe la función merge. Con merge(), podemos utilizar los campos en común entre las tablas (en este ejemplo: CIUDAD y DEPARTAMENTO) para unificar la información en un solo lugar.

In [46]:
# Dicho esto veamos cómo funciona. Empecemos por separar en dos la tabla que tenemos.
# Es decir, de forma artificial vamos a hacer de cuenta que tenemos dos tablas

columnas_1 = ["NOMBRE AEROPUERTO", "CIUDAD", "Pasajeros Salidos - 2019", "Pasajeros Llegados - 2019"]
df_1 = df[columnas_1]
df_1.head()

# Esta es nuestra tabla artificial 1

NOMBRE AEROPUERTO       CIUDAD  Pasajeros Salidos - 2019  \
0                           EL ENCANTO   EL ENCANTO                      62.0   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA                     788.0   
2                           LA PEDRERA   LA PEDRERA                    2517.0   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA                  166039.0   
4                             TARAPACA      LETICIA                     758.0   

   Pasajeros Llegados - 2019  
0                       55.0  
1                      859.0  
2                     2570.0  
3                   167011.0  
4                      701.0

In [47]:
# Miremos la tabla artificial 2

columnas_2 = ["NOMBRE AEROPUERTO", "CIUDAD", "Pasajeros Salidos - 2018", "Pasajeros Llegados - 2018"]
df_2 = df[columnas_2]
df_2.head()

NOMBRE AEROPUERTO       CIUDAD  Pasajeros Salidos - 2018  \
0                           EL ENCANTO   EL ENCANTO                      87.0   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA                    1167.0   
2                           LA PEDRERA   LA PEDRERA                    2731.0   
3         LETICIA-ALFREDO VASQUEZ COBO      LETICIA                  134304.0   
4                             TARAPACA      LETICIA                    1164.0   

   Pasajeros Llegados - 2018  
0                       76.0  
1                     1278.0  
2                     2927.0  
3                   135994.0  
4                     1242.0

In [48]:
# Entonces tenemos df_1 y df_2. Veamos cómo juntarlos con merge:
# primero va la tabla a la que vamos a agregar los datos nuevos
df_1.merge(
           right=df_2, # Esta es la segunda tabla que queremos juntar
           left_on=["NOMBRE AEROPUERTO", "CIUDAD"], # Las columnas que usaremos en la primera tabla para juntar la info
           right_on=["NOMBRE AEROPUERTO", "CIUDAD"], # Las columnas que usaremos en la segunda tabla para juntar la info
           how="left" # Este es el tipo de cruce. 'left' significa que 2 se pone en 1; 'right' que 1 se pone en 2
                      # hay otros tipos de cruces de los que vamos a hablar en el BootCamp
          )

NOMBRE AEROPUERTO                CIUDAD  \
0                             EL ENCANTO            EL ENCANTO   
1    LA CHORRERA - VIRGILIO BARCO VARGAS           LA CHORRERA   
2                             LA PEDRERA            LA PEDRERA   
3           LETICIA-ALFREDO VASQUEZ COBO               LETICIA   
4                               TARAPACA               LETICIA   
..                                   ...                   ...   
826                        SANTA BARBARA               CUMARAL   
827                     YATAROS DE MELUA          PUERTO LOPEZ   
828                     LA CABA?A - META  SAN CARLOS DE GUAROA   
829                       AEROSPORT.C.A.               FLANDES   
830               INGENIO MAYAGUEZ (C.A)            CANDELARIA   

     Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
0                        62.0                       55.0   
1                       788.0                      859.0   
2                      2517.0                     2570.0   
3                    166039.0                   167011.0   
4                       758.0                      701.0   
..                        ...                        ...   
826                       0.0                        0.0   
827                       0.0                        0.0   
828                       0.0                        0.0   
829                       0.0                        0.0   
830                       0.0                        0.0   

     Pasajeros Salidos - 2018  Pasajeros Llegados - 2018  
0                        87.0                       76.0  
1                      1167.0                     1278.0  
2                      2731.0                     2927.0  
3                    134304.0                   135994.0  
4                      1164.0                     1242.0  
..                        ...                        ...  
826                       0.0                        0.0  
827                       0.0                        0.0  
828                       0.0                        0.0  
829                       0.0                        0.0  
830                       0.0                        0.0  

[831 rows x 6 columns]

In [78]:
df_2.groupby(by=["NOMBRE AEROPUERTO", "CIUDAD"]).count().sort_values(by="Pasajeros Salidos - 2018")

Pasajeros Salidos - 2018  \
NOMBRE AEROPUERTO              CIUDAD                                         
 PORTO ALEGRE    SALGADO FILHE PORTO ALEGRE                               1   
MURIBA                         MANI                                       1   
MURINDO                        MURINDO                                    1   
MUTATA                         MUTATA                                     1   
MUZO                           MUZO                                       1   
...                                                                     ...   
CACHIPORRO                     MITU                                       2   
PUERTO WILCHES                 PUERTO WILCHES                             2   
CURUMANI                       CURUMANI                                   2   
LAS BRISAS                     VILLAVICENCIO                              2   
SANTA ROSA DEL SUR             SANTA ROSA DEL SUR                         3   

                                                   Pasajeros Llegados - 2018  
NOMBRE AEROPUERTO              CIUDAD                                         
 PORTO ALEGRE    SALGADO FILHE PORTO ALEGRE                                1  
MURIBA                         MANI                                        1  
MURINDO                        MURINDO                                     1  
MUTATA                         MUTATA                                      1  
MUZO                           MUZO                                        1  
...                                                                      ...  
CACHIPORRO                     MITU                                        2  
PUERTO WILCHES                 PUERTO WILCHES                              2  
CURUMANI                       CURUMANI                                    2  
LAS BRISAS                     VILLAVICENCIO                               2  
SANTA ROSA DEL SUR             SANTA ROSA DEL SUR                          3  

[781 rows x 2 columns]

In [80]:
df[df.CIUDAD=="SANTA ROSA DEL SUR"]

NOMBRE AEROPUERTO              CIUDAD DEPARTAMENTO  \
34   SANTA ROSA DEL SUR  SANTA ROSA DEL SUR      BOLIVAR   
527  SANTA ROSA DEL SUR  SANTA ROSA DEL SUR      BOLIVAR   
614  SANTA ROSA DEL SUR  SANTA ROSA DEL SUR      BOLIVAR   

     Pasajeros Salidos - 2019  Pasajeros Llegados - 2019  \
34                      913.0                      873.0   
527                       0.0                        0.0   
614                       0.0                        0.0   

     Carga y Correo Salida (Kg) - 2019  Carga y Correo Llegados (kg) - 2019  \
34                              2801.0                               3153.0   
527                                0.0                                  0.0   
614                                0.0                                  0.0   

     Número Vuelos Salidos - 2019  Número Vuelos Llegados - 2019  \
34                          432.0                          434.0   
527                           0.0                            0.0   
614                           0.0                            0.0   

     Pasajeros Salidos - 2018  ...  Carga y Correo Llegados (kg) - 2005  \
34                      859.0  ...                                  0.0   
527                       0.0  ...                               2344.0   
614                       0.0  ...                                  0.0   

     Número Vuelos Salidos - 2005  Número Vuelos Llegados - 2005  \
34                            0.0                            0.0   
527                         255.0                          256.0   
614                           0.0                            0.0   

     Pasajeros Salidos - 2004  Pasajeros Llegados - 2004  \
34                        0.0                        0.0   
527                    1213.0                     1081.0   
614                       0.0                        0.0   

     Carga y Correo Salida (Kg) - 2004  Carga y Correo Llegados (kg) - 2004  \
34                                 0.0                                  0.0   
527                             1600.0                               4633.0   
614                                0.0                                  0.0   

     Número Vuelos Salidos - 2004  Número Vuelos Llegados - 2004  \
34                            0.0                            0.0   
527                         332.0                          336.0   
614                           0.0                            0.0   

     nueva_columna  
34               0  
527              0  
614              0  

[3 rows x 100 columns]

¡Y listo! Ahora tenemos toda la información en un mismo lugar

## Ejercicio 5

Explore las tablas df_3 y df_4 y cruce la información en una nueva tabla llamada df_5

In [58]:
# Puede ignorar esto, pero deben correrlo una vez

columnas = ["NOMBRE AEROPUERTO", "CIUDAD", "Número Vuelos Salidos - 2018", "Número Vuelos Salidos - 2018"]
df_3 = df[columnas]
columnas = ["NOMBRE AEROPUERTO", "CIUDAD", "Número Vuelos Salidos - 2019", "Número Vuelos Salidos - 2019"]
random.shuffle(columnas)
df_4 = df[columnas].rename(columns={"NOMBRE AEROPUERTO":"aeropuerto",
                                    "CIUDAD":"CiTy"})


In [59]:
df_3.head(2)

NOMBRE AEROPUERTO       CIUDAD  \
0                           EL ENCANTO   EL ENCANTO   
1  LA CHORRERA - VIRGILIO BARCO VARGAS  LA CHORRERA   

   Número Vuelos Salidos - 2018  Número Vuelos Salidos - 2018  
0                          38.0                          38.0  
1                         205.0                         205.0

In [60]:
df_4.head(2)

Número Vuelos Salidos - 2019                           aeropuerto  \
0                          31.0                           EL ENCANTO   
1                         136.0  LA CHORRERA - VIRGILIO BARCO VARGAS   

          CiTy  Número Vuelos Salidos - 2019  
0   EL ENCANTO                          31.0  
1  LA CHORRERA                         136.0

In [66]:
df_3.shape

(797, 4)

In [ ]:
# Acá va su código

df_5 = # Acá va el merge

## Ejercicio 6

Utilizando la información del ejercicio anterior, responda las siguientes preguntas:

* ¿Cuál fue el departamento con mayor número de vuelos salidos promedio en el 2019?
* ¿Cuál fue el departamento con menor número de vuelos salidos absolutos en el 2018?

In [63]:
df_5 = df_3.merge(df_4,
                  left_on=["NOMBRE AEROPUERTO", "CIUDAD"],
                  right_on=["aeropuerto","CiTy"],
                  how='left')

In [64]:
df_5.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 831 entries, 0 to 830
Data columns (total 8 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   NOMBRE AEROPUERTO             831 non-null    object 
 1   CIUDAD                        831 non-null    object 
 2   Número Vuelos Salidos - 2018  831 non-null    float64
 3   Número Vuelos Salidos - 2018  831 non-null    float64
 4   Número Vuelos Salidos - 2019  831 non-null    float64
 5   aeropuerto                    831 non-null    object 
 6   CiTy                          831 non-null    object 
 7   Número Vuelos Salidos - 2019  831 non-null    float64
dtypes: float64(4), object(4)
memory usage: 58.4+ KB
